pennylane/resources/factorization.py

```diff
@@ -0,0 +1,127 @@
+
+    .. math::
+
+        H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} T_{pq} a_{p,\alpha}^{\dagger}
+        a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
+        V_{pqrs} a_{p, \alpha}^{\dagger} a_{q, \alpha} a_{r, \beta}^{\dagger} a_{s, \beta}.
+
+    with
+
+    .. math::
+
+        T_{pq} = h_{ij} - \frac{1}{2} \sum_s h_{pssq}.
+
+
+    and :math:`V` is the two-electron tensor in chemist notation.
+
+    The objective of the factorization is to find a set of symmetric matrices, :math:`L`, such that
+
+    .. math::
+
+           V_{ijkl} = \sum_r L_{ij}^{(r)} L_{kl}^{(r) T}.
+
+    with the rank :math:`r \in \mathcal{O}(n)`. The matrices :math:`L` are further diagonalized
+    and truncated in a second level of factorization.
+
+    The algorithm has the following steps
+    [`arXiv:1902.02134 <https://arxiv.org/pdf/1902.02134.pdf>`_].
+
+    1. Matricize the :math:`n \times n \times n \times n` two-electron tensor to a \
+        :math:`n^2 \times n^2` matrix where n is the number of orbitals.
+
+    2. Diagonalize the resulting matrix and keep the :math:`r` eigenvectors that have \
+        corresponding eigenvalues larger than a threshold.
+
+    3. Reshape the selected eigenvectors to :math:`n \times n` matrices.
+
+    4. Diagonalize the :math:`n \times n` matrices and keep those that the norm of their \
+        eigenvalues is larger than a threshold.
+
+    Args:
+        two (array[array[float]]): the two-electron repulsion tensor in the molecular orbital basis
+            arranged in chemist notation [11|22]
+        tol (float): cutoff value for discarding the negligible factors
+
+    Returns:
+        tuple(array[float]): array of symmetric matrices (factors) approximating the two-electron
+        tensor, eigenvalues of the generated factors, eigenvectors of the generated factors
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.74, 0.0, 0.0]], requires_grad = False) / 0.5291772
+    >>> mol = qml.qchem.Molecule(symbols, geometry)
+    >>> core, one, two = qml.qchem.electron_integrals(mol)()
+    >>> two = np.swapaxes(two, 1, 3) # convert to chemist's notation
+    >>> l, w, v = factorize(two, 1e-5)
+    >>> print(l)
+    [[[ 1.06723440e-01  9.73575768e-15]
+      [ 8.36288956e-15 -1.04898533e-01]]
+     [[-2.20945401e-13 -4.25688222e-01]
+      [-4.25688222e-01 -2.98228790e-13]]
+     [[-8.14472856e-01  5.01669019e-13]
+      [ 5.01689072e-13 -8.28642140e-01]]]
+    """
+    n = two.shape[0]
+    two = two.reshape(n * n, n * n)
+
+    eigvals, eigvecs = np.linalg.eigh(two)
+    eigvals = np.array([val for val in eigvals if abs(val) > tol])
+    eigvecs = eigvecs[:, -len(eigvals) :]
+
+    vectors = eigvecs @ np.diag(np.sqrt(abs(eigvals)))
+
+    factors = np.array([vectors.reshape(n, n, len(eigvals))[:, :, r] for r in range(len(eigvals))])
+
+    eigvals, eigvecs = np.linalg.eigh(factors)
```

Comment:
should we use different variable names to avoid confusion and overwriting? Here we're reusing `eigvecs` and `eigvals`